frontend/dockerfile/dockerfile2llb/outline.go

```diff
@@ -0,0 +1,210 @@
+package dockerfile2llb
+
+import (
+	"sort"
+
+	"github.com/moby/buildkit/frontend/dockerfile/instructions"
+	"github.com/moby/buildkit/frontend/dockerfile/parser"
+	"github.com/moby/buildkit/frontend/subrequests/outline"
+	pb "github.com/moby/buildkit/solver/pb"
+)
+
+type outlineCapture struct {
+	allArgs  map[string]argInfo
+	usedArgs map[string]struct{}
+	secrets  map[string]secretInfo
+	ssh      map[string]sshInfo
+}
+
+type argInfo struct {
+	value      string
+	definition instructions.KeyValuePairOptional
+	deps       map[string]struct{}
+	location   []parser.Range
+}
+
+type secretInfo struct {
+	required bool
+	location []parser.Range
+}
+
+type sshInfo struct {
+	required bool
+	location []parser.Range
+}
+
+func newOutlineCapture() outlineCapture {
+	return outlineCapture{
+		allArgs:  map[string]argInfo{},
+		usedArgs: map[string]struct{}{},
+		secrets:  map[string]secretInfo{},
+		ssh:      map[string]sshInfo{},
+	}
+}
+
+func (o outlineCapture) clone() outlineCapture {
+	allArgs := map[string]argInfo{}
+	for k, v := range o.allArgs {
+		allArgs[k] = v
+	}
+	usedArgs := map[string]struct{}{}
+	for k := range o.usedArgs {
+		usedArgs[k] = struct{}{}
+	}
+	secrets := map[string]secretInfo{}
+	for k, v := range o.secrets {
+		secrets[k] = v
+	}
+	ssh := map[string]sshInfo{}
+	for k, v := range o.ssh {
+		ssh[k] = v
+	}
+	return outlineCapture{
+		allArgs:  allArgs,
+		usedArgs: usedArgs,
+		secrets:  secrets,
+		ssh:      ssh,
+	}
+}
+
+func (o outlineCapture) markAllUsed(in map[string]struct{}) {
+	for k := range in {
+		if a, ok := o.allArgs[k]; ok {
+			o.markAllUsed(a.deps)
+		}
+		o.usedArgs[k] = struct{}{}
+	}
+}
+
+func (ds *dispatchState) args(visited map[string]struct{}) []outline.Arg {
+	ds.outline.markAllUsed(ds.outline.usedArgs)
+
+	args := make([]outline.Arg, 0, len(ds.outline.usedArgs))
+	for k := range ds.outline.usedArgs {
+		if a, ok := ds.outline.allArgs[k]; ok {
+			if _, ok := visited[k]; !ok {
+				args = append(args, outline.Arg{
+					Name:        a.definition.Key,
+					Value:       a.value,
+					Description: a.definition.Comment,
```

Comment:
I wonder if the comments used as description for args and stages should have a special markup to avoid unwanted behavior like https://github.com/crazy-max/moby/blob/82ede62397d3de933e79483f84d3a8092587485f/Dockerfile#L237-L239

I have this use case for example where I want to explicitly tell what stages the user should care about: https://github.com/crazy-max/moby/blob/82ede62397d3de933e79483f84d3a8092587485f/Dockerfile#L618-L643

Reply:
It already does this